rbac/management/group/definer.py

```diff
@@ -144,31 +149,41 @@ def add_roles(group, roles_or_role_ids, tenant, user=None):
                     "into groups."
                 )
                 raise serializers.ValidationError({key: _(message)})
+
         # Only add the role if it was not attached
-        if not system_policy.roles.filter(pk=role.pk).exists():
-            system_policy.roles.add(role)
+        if system_policy.roles.filter(pk=role.pk).exists():
+            continue
 
-            # Send notifications
-            group_role_change_notification_handler(user, group, role, "added")
+        system_policy.roles.add(role)
 
+        dual_write_handler = RelationApiDualWriteGroupHandler(group, ReplicationEventType.ASSIGN_ROLE, [])
+        dual_write_handler.replicate_added_role(role)
+        # Send notifications
+        group_role_change_notification_handler(user, group, role, "added")
 
+
+@transaction.atomic
 def remove_roles(group, roles_or_role_ids, tenant, user=None):
     """Process list of roles and remove them from the group."""
-    if not isinstance(roles_or_role_ids, QuerySet):
-        # If given an iterable of UUIDs, get the corresponding objects
-        roles = Role.objects.filter(uuid__in=roles_or_role_ids)
-    else:
-        roles = roles_or_role_ids
-    role_names = list(roles.values_list("name", flat=True))
-
+    roles = _roles_by_query_or_ids(roles_or_role_ids)
```

Comment:
I didn't realize I slightly changed the query in this case from the prior code. Previously it didn't bother filtering the tenant. I assume this was because if you're removing a role, it must be because the role was allowed to be there in the first place, so it doesn't really do anything if the role happens to be a role from another tenant. It would be a no-op. I guess it doesn't matter, but just pointing this out in case anyone thinks there are any issues with this.

Reply:
I mispoke – the query is the same.